File: JavaScript/1026-Maximum Difference Between Node and Ancestor.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxAncestorDiff = function(root) {
    maxValue = Number.NEGATIVE_INFINITY;
    minValue = Number.POSITIVE_INFINITY;
    res = 0
    let dfs = (root) => {
        if (!root){
            return [maxValue, minValue];
        }
        let [leftMax, leftMin] = dfs(root.left);
        let [rightMax, rightMin] = dfs(root.right);

        let currMax = Math.max(root.val,leftMax, rightMax);
        let currMin = Math.min(root.val, leftMin, rightMin);     
        res = Math.max(res, Math.abs(root.val-currMax), Math.abs(root.val-currMin));
        return [currMax, currMin];
    }
    dfs(root);
    return res;
};

File: C++/1026-Maximum Difference Between Node and Ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 0;
    int maxAncestorDiff(TreeNode* root) {
        int curr = root->val;
        helper(root, curr, curr);
        return res;
    }
    
    void helper(TreeNode* node, int maxVal, int minVal) {
        int curr = node->val;
        maxVal = max(maxVal, curr);
        minVal = min(minVal, curr);
        res = max(maxVal-minVal, res);
        if (node->left) helper(node->left, maxVal, minVal);
        if (node->right) helper(node->right, maxVal, minVal);
    }
};